### PR TITLE
[W5.6a][W12-1] Lim Jit Wei

### PR DESCRIPTION
**Pull request by jitwei98**
Extract common attributes and behaviours from Address, Email and Phone classes to Contact class

### CODE REVIEW

File: src/seedu/addressbook/data/person/Address.java [hunk 1]
```diff
@@ -4,56 +4,32 @@
...
-    public static boolean isValidAddress(String test) {
-        return test.matches(ADDRESS_VALIDATION_REGEX);
-    }
-
-    @Override
-    public String toString() {
-        return value;
-    }
-
     @Override
```
Comment (jitwei98) [inline src/seedu/addressbook/data/person/Address.java]: Still haven't figured out how to pass the class name 'Address' or 'Email' or 'Phone' as parameters into a method

File: src/seedu/addressbook/data/person/Address.java [hunk 2]
```diff
@@ -4,56 +4,32 @@
...
     public boolean equals(Object other) {
```
Comment (jitwei98) [inline src/seedu/addressbook/data/person/Address.java]: Still haven't figured out how to pass the class name 'Address' or 'Email' or 'Phone' as parameters into a method

File: src/seedu/addressbook/data/person/Address.java [hunk 3]
```diff
@@ -4,56 +4,32 @@
...
         return other == this // short circuit if same object
                 || (other instanceof Address // instanceof handles nulls
                 && this.value.equals(((Address) other).value)); // state check
```
Comment (jitwei98) [inline src/seedu/addressbook/data/person/Address.java:33]: Still haven't figured out how to pass the class name '**Address**' or '**Email**' or '**Phone**' as parameters into a method

File: src/seedu/addressbook/data/person/Address.java [hunk 4]
```diff
@@ -4,56 +4,32 @@
...
      * Validates given address.
      *
      * @throws IllegalValueException if given address string is invalid.
      */
     public Address(String address, boolean isPrivate) throws IllegalValueException {
+        super(isPrivate);
         String trimmedAddress = address.trim();
-        this.isPrivate = isPrivate;
-        if (!isValidAddress(trimmedAddress)) {
+        if (!isValid(trimmedAddress, ADDRESS_VALIDATION_REGEX)) {
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java:23]: Think how you can push the validation to the super class as well.

File: src/seedu/addressbook/data/person/Contact.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Override
+    public String toString() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof Contact // instanceof handles nulls
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Contact.java:30]: This could be a potential problem. You may end up equating email is a phone number. Think why?

**Review by okkhoy — APPROVED** [review]
You can consider updating the tests based on the changes here.

**Comment by okkhoy** [comment]
@jitwei98 
Some comments added. Please take note and close the pull request.

**Pull request closed without merging**